homeassistant/components/automation/trace.py

```diff
@@ -204,3 +206,23 @@ def get_debug_traces(hass, summary=False):
         )
 
     return traces
+
+
+class TraceJSONEncoder(json.JSONEncoder):
+    """JSONEncoder that supports Home Assistant objects and falls back to repr(o)."""
+
+    def default(self, o: Any) -> Any:
+        """Convert certain objects.
+
+        Fall back to repr(o).
+        """
+        if isinstance(o, datetime):
+            return o.isoformat()
+        if isinstance(o, timedelta):
+            return str(o)
+        if isinstance(o, set):
+            return list(o)
+        if hasattr(o, "as_dict"):
+            return o.as_dict()
```

Comment:
I think that we shouldn't implement any of this logic. Instead we should inherit from the HA JSONEncoder, and add the `__type` on top when it fails.